source/s3_buffer_pool.c

```diff
@@ -0,0 +1,225 @@
+        aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+    }
+
+    aws_array_list_clean_up(&buffer_pool->blocks);
+
+    aws_mutex_clean_up(&buffer_pool->mutex);
+    struct aws_allocator *base = buffer_pool->base_allocator;
+    aws_mem_release(base, buffer_pool);
+}
+
+void aws_s3_buffer_pool_trim(struct aws_s3_buffer_pool *buffer_pool) {
+    size_t swap_count = 0;
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        if (block->block_ptr == 0) {
+            size_t swap_loc = aws_array_list_length(&buffer_pool->blocks) - 1 - swap_count;
+            if (swap_loc == i) {
+                break;
+            } else {
+                aws_array_list_swap(&buffer_pool->blocks, i, swap_loc);
+            }
+        }
+    }
+
+    for (size_t i = 0; i < swap_count; ++i) {
+        buffer_pool->current_mem_usage -= buffer_pool->block_size;
+        aws_array_list_pop_back(&buffer_pool->blocks);
+    }
+}
+
+static struct aws_s3_pooled_buffer s_primary_acquire(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+    uint8_t *alloc_ptr = NULL;
+    aws_mutex_lock(&buffer_pool->mutex);
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        if (block->alloc_count == 0) {
+            block->offset_ptr = block->block_ptr + size;
+            alloc_ptr = block->block_ptr;
+            block->alloc_count += 1;
+            break;
+        }
+
+        if (block->offset_ptr + size < block->block_ptr + block->block_size) {
+            alloc_ptr = block->offset_ptr;
+            block->offset_ptr += size;
+            block->alloc_count += 1;
+            break;
+        }
+    }
+
+    if (alloc_ptr == NULL && (buffer_pool->max_mem_usage ||
+                              buffer_pool->current_mem_usage + buffer_pool->block_size <= buffer_pool->max_mem_usage)) {
+        struct s3_buffer_pool_block block;
+        block.alloc_count = 1;
+        block.block_ptr = aws_mem_acquire(buffer_pool->base_allocator, buffer_pool->block_size);
+        block.offset_ptr = block.block_ptr + size;
+        block.block_size = buffer_pool->block_size;
+        aws_array_list_push_back(&buffer_pool->blocks, &block);
+        buffer_pool->current_mem_usage += buffer_pool->block_size;
+        alloc_ptr = block.block_ptr;
+    }
+
+    aws_mutex_unlock(&buffer_pool->mutex);
+
+    return (struct aws_s3_pooled_buffer){.ptr = alloc_ptr, .size = size};
+}
+
+struct aws_s3_pooled_buffer aws_s3_buffer_pool_acquire_buffer(struct aws_s3_buffer_pool *buffer_pool, size_t size) {
+    AWS_PRECONDITION(buffer_pool);
+
+    if (size < buffer_pool->block_size) {
+        return s_primary_acquire(buffer_pool, size);
```

Comment:
the locking in this function seems error prone. Maybe lock in this function rather than s_primary_acquire?